### DISCUSSION TITLE
TSQL question

### DISCUSSION THREAD

**Discussion by odeddror** [community/community#67323] [Copilot Conversations] [answered]
### Select Topic Area

Question

### Body

Hi there,
I have expected result as a view but I wanted to ask the right question in TSQL, here is the schenario

/*
The query grabs the first time for each ID (the first SELECT, before the UNION ALL) 
and then recurses, finding the next row for each ID which is at least 15 minutes appart.
*/
CREATE TABLE example(
       [ID] [varchar](9) NULL,
       [DTTM] [datetime] NULL,
       [RN] Int null,
       [COUNTFLAG] int
) ON [PRIMARY] 

Insert into [example] (ID,DTTM) values 
('123456789','2017-10-05 08:00:00.000'), -- Expected Result
('123456789','2017-10-05 08:05:00.000'), 
('123456789','2017-10-05 08:07:00.000'),
('123456789','2017-10-05 08:15:00.000'), -- Expected Result
('123456789','2017-10-05 08:25:00.000'),
('123456789','2017-10-05 10:12:00.000') , -- Expected Result
('123456789','2017-10-05 10:26:00.000'),
('123456789','2017-10-05 10:32:00.000'),  -- Expected Result
('123456789','2017-10-05 10:33:00.000'),
('123456789','2017-10-05 10:34:00.000'),
('123456789','2017-10-05 10:35:00.000'),
('123456789','2017-10-05 10:36:00.000'),
('123456789','2017-10-05 10:37:00.000'),
('123456789','2017-10-05 10:38:00.000'),
('123456789','2017-10-05 10:39:00.000'),
('123456789','2017-10-05 10:40:00.000'),
('123456789','2017-10-05 10:41:00.000'),
('123456789','2017-10-05 10:42:00.000'),
('123456789','2017-10-05 10:43:00.000'),
('123456789','2017-10-05 10:44:00.000'),
('123456789','2017-10-05 10:45:00.000'),
('123456789','2017-10-05 10:46:00.000'),
('123456789','2017-10-05 10:47:00.000'),  -- Expected Result
('123456789','2017-10-05 10:48:00.000')

Here is the answer
WITH times AS (
    SELECT e.ID, MIN(e.DTTM) AS DTTM
    FROM example e
    GROUP BY e.ID

    UNION ALL

    SELECT r.ID, r.DTTM
    FROM (
        SELECT e.ID, e.DTTM,
            ROW_NUMBER() OVER (PARTITION BY e.ID ORDER BY e.DTTM) AS rn
        FROM example e
        INNER JOIN times t
            ON e.ID = t.ID AND e.DTTM >= DATEADD(mi, 15, t.DTTM)
    ) r
    WHERE r.rn = 1
)
SELECT top 100 percent ID,DTTM
FROM times
ORDER BY ID, DTTM

How do I ask CoPilot a question to produce this code?
I'm looking for how to ask good question
Thanks,
Oded Dror

**Comment by mkare** [comment, marked as answer]
Here's an example of how you might ask CoPilot to generate the SQL code for your scenario:

I have a table named "example" with columns "ID" and "DTTM." I need to retrieve the first occurrence for each "ID" and then find subsequent occurrences that are at least 15 minutes apart from the previous one. Can you generate SQL code using Common Table Expressions (CTEs) or subqueries to achieve this?

Sample Data:
- Table: "example"
  - Columns: "ID" (varchar), "DTTM" (datetime)

Please provide SQL code that accomplishes this task.

By providing a clear problem statement, specifying your data and requirements, and mentioning your preferred approach (in this case, CTEs or subqueries), you'll increase the likelihood of getting a helpful code suggestion from CoPilot.
